Dev guide: Port to markbind, add onboarding info

kkangs0226 opened the pull request:
*Edited on 25/04/2021
# Summary

Fixes https://github.com/CATcher-org/CATcher/issues/667

## Changes made:
- Created a Onboarding Guide for new developers
- Added introduction to the tools and frameworks used in CATcher (I referred to our presentation slides for certain parts, credits to @dingyuchen, as well as my own Knowledge page on the 3281 website for this)
- Merged existing developer notes (https://github.com/CATcher-org/CATcher/blob/master/docs/developer-notes.md) content into Onboarding Guide 


## To-do:
- Add tutorials specific to CATcher in another PR


### Proposed commit message:
```
Developers' guide: Port to markbind, add onboarding info

The current Developers' Guide lacks introduction to basic tools
and frameworks used in CATcher.

We can incorporate the content of the existing developer guide
into the new Developers' Guide to make the learning process more
convenient for developers who are new to CATcher.

We can also add headers and site navigation for Developers' guide
to link to CATcher's Markbind site. 

Let's
* Add onboarding information into the Developers' guide for 
new developers
* Merge the content of the existing Developer's Notes to this 
Developer's Guide for easier viewing
* Add introduction for the tools and frameworks used in CATcher
* Add headers and site navigation for Developers' guide
```

ptvrajsk commented on dg/onboarding-guide.md:
```diff
@@ -0,0 +1,292 @@
+          keywords="injectable" threshold="0.5" answer="@Injectable()">
+  <div slot="hint">
+
+  Read up on the [Angular Guide - Services and Dependency Injection](https://angular.io/guide/architecture#services-and-dependency-injection) for more details on services and dependency injection.
```
For Angular shall we get the new devs to walkthrough Angular's tutorial to get an understanding of it. https://angular.io/tutorial
So like do the tutorial then try the question?

It was what we did when I first joined the project to learn Angular 😆

kkangs0226 replied:
Yes sure, I can add that into the docs! The tutorials we give them will likely not be enough😭

ptvrajsk replied:
Yeah but thats okay though I think as long as we can point them to some possible resources to refer to. Can't expect this page to be a "module" on web development 😆

kkangs0226 replied:
Yeah, I agree! Just a few will be enough imo, just some checkpoints for students to check their understanding aft visiting the resources (I think that's what Prof. Damith intended for us to do!)

anubh-v commented:
I'm not sure we should add quizzes in this guide.
If we do, we would have the additional work of updating them when there are new advancements in these libraries / frameworks (e.g. changes in implementation of Observables, new best practices in Jasmine / Angular).

There are specific resources on the web meant for learning these topics. Would it be sufficient to point devs to these resources?

kkangs0226 commented:
> I'm not sure we should add quizzes in this guide.
> If we do, we would have the additional work of updating them when there are new advancements in these libraries / frameworks (e.g. changes in implementation of Observables, new best practices in Jasmine / Angular).
> 
> There are specific resources on the web for learning these topics. Would it be sufficient to point devs to these resources?

Thanks for pointing it out, I never thought so far 😭 The idea of having tutorials was initially from Prof Damith, we should ask for Prof's thoughts on this @damithc ! Not sure if these "quizzes" were what Prof meant by tutorials. 

For me, I think having these tutorials specific to CATcher (so maybe the quizzes should be more relevant to CATcher's context, so it focuses more on the understanding the structure of our application and not really on specific tools/functions we use) will be good, it will definitely help the new devs in checking their understanding of our app. They will learn the technicalities and tools from the resources like you mentioned, and our "quizzes" will test their understanding of CATcher specifically. Again, we would have to update these quizzes if we make major changes to our structures, so the problem that you mentioned would still remain. 

What do you think @anubh-v @damithc @CATcher-org/developers ?

damithc commented:
> For me, I think having these tutorials specific to CATcher (so maybe the quizzes should be more relevant to CATcher's context, so it focuses more on the understanding the structure of our application and not really on specific tools/functions we use) will be good, it will definitely help the new devs in checking their understanding of our app. They will learn the technicalities and tools from the resources like you mentioned, and our "quizzes" will test their understanding of CATcher specifically. Again, we would have to update these quizzes if we make major changes to our structures, so the problem that you mentioned would still remain.

Yes, make them specific to CATcher and based on important things that are unlikely to change often. Don't need too many questions either.

kkangs0226 commented:
> > For me, I think having these tutorials specific to CATcher (so maybe the quizzes should be more relevant to CATcher's context, so it focuses more on the understanding the structure of our application and not really on specific tools/functions we use) will be good, it will definitely help the new devs in checking their understanding of our app. They will learn the technicalities and tools from the resources like you mentioned, and our "quizzes" will test their understanding of CATcher specifically. Again, we would have to update these quizzes if we make major changes to our structures, so the problem that you mentioned would still remain.
> 
> Yes, make them specific to CATcher and based on important things that are unlikely to change often. Don't need too many questions either.

Alright, will do!

anubh-v commented:
We can also add in the tutorials / quizzes in a separate PR, and just setup an initial guide in this PR.

kkangs0226 commented:
> We can also add in the tutorials / quizzes in a separate PR, and just setup an initial guide in this PR.

I have removed the existing quizzes, will make a separate PR to add in the tutorials after merging the initial guide in!

anubh-v commented:
Possible to add in a header to the onboarding guide page (just like the user guide) ?
It'll help to link to other pages of the site (home page, user guide) ?

kkangs0226 commented:
> Possible to add in a header to the onboarding guide page (just like the user guide) ?
> It'll help to link to other pages of the site (home page, user guide) ?

I've added the headers, and formatted it such that it follows the style of the User Guide! 
I also renamed the Onboarding Guide to Developers' Guide, after merging the DG content into the Onboarding Guide. Not sure if we should stick to Onboarding Guide or Developers' Guide, Developers' Guide might be more familiar to students.

anubh-v commented on dg/workflow.md:
```diff
@@ -0,0 +1,53 @@
+
+## Testing
+
+We use [Scuri](https://github.com/gparlakov/scuri) to generate spec files for unit tests.
```
I think we can remove the info on generating spec files with Scuri, as we haven't been using it much for our tests.

anubh-v commented on dg/tools-frameworks.md:
```diff
@@ -0,0 +1,139 @@
+
+GraphQL is a query language for APIs.
+
+In CATcher, it is to communicate with the Github backend. It allows us to define a schema for the data that we receive from github, and also performs validation for the data we reqeust for, such that the data we receive from Github is exactly what we need, nothing more, nothing less. The extra layer of GraphQL is one that makes the process of data retrieval in CATcher more streamlined.
```
Can mention that the GraphQL queries used by CATcher are defined under the `graphql` folder in the codebase.

anubh-v commented on dg/tests.md:
````diff
@@ -0,0 +1,166 @@
+
+Code snippet of a component test in CATcher
+```
+describe('AssigneeComponent', () => {
````
I think we should minimise usage of code snippets that can change over time.
Moreover, this snippet doesn't show the actual tests (just the component setup).
How about we just link to the test file itself?

kkangs0226 replied:
That sounds better!

anubh-v commented on dg/tests.md:
```diff
@@ -0,0 +1,166 @@
+
+## Jasmine Style Guide
+
+Jasmine is a behavior-driven development framework specific for testing JavaScript code. We follow the Jasmine Style Guide loosely for our tests.
```
Looks like you missed out the link to the [style guide](https://github.com/CareMessagePlatform/jasmine-styleguide)

kkangs0226 replied:
Thanks for pointing it out!

anubh-v commented on dg/tests.md:
```diff
@@ -0,0 +1,166 @@
+
+### Test Suite
+
+Jasmine testing involves the test suite, which is started by calling the global Jasmine function `describe`. The test suite may consist of several tests (specs) within itself, which is done by calling `it`. One main guideline is that a `describe` block should be created for each method / scenario under test, and an `it` block should be created for each property being verified.
```
We can keep the following sentence (as it's a style guideline):
`One main guideline is that a describe block should be created for each method / scenario under test, and an `it` block should be created for each property being verified.`

Let's move the sentence to the paragraph where the style guide is introduced:

`
Jasmine is a behavior-driven development framework specific for testing JavaScript code. We follow the Jasmine Style Guide loosely for our tests. One main guideline is that a describe block should be created for each method / scenario under test, and an `it` block should be created for each property being verified.`
`

anubh-v commented on dg/tests.md:
```diff
@@ -0,0 +1,166 @@
+
+Jasmine is a behavior-driven development framework specific for testing JavaScript code. We follow the Jasmine Style Guide loosely for our tests.
+
+### Test Suite
```
In my view, there is no need to describe how Jasmine or its various features are used (unless there is some non-trivial information specific to CATcher, as seen in the case of the E2E tests).

For each tool used by our project, it should be sufficient to explain *what* the tool is / *why* is it needed in the project.

If we include detailed info on how to use the tool / how it works, we'll simply be replicating information already provided on the web by experts. We'll also have to ensure that the info stays updated.

Hence, let's remove the explanation provided here for Jasmine (the `Test Suite` and `Spy` sections).

kkangs0226 replied:
Sure, I have removed it!

anubh-v commented:
> I also renamed the Onboarding Guide to Developers' Guide, after merging the DG content into the Onboarding Guide. Not sure if we should stick to Onboarding Guide or Developers' Guide, Developers' Guide might be more familiar to students.

Sure, we can name it Developers' Guide.

anubh-v commented on dg/tools-frameworks.md:
```diff
@@ -0,0 +1,141 @@
+Electron is a tool that allows developers to build cross-platform desktop applications with Javascript, HTML, and CSS. It basically allows us to "convert" the web version of CATcher into a desktop application through the APIs provided by Electron.
+
+### How Electron Works
+Electron uses the module `ipcMain` to communicate to renderer processes from the main process. It sends out and listen for events, enabling inter-process communication within the CATcher.
```
We can improve this section by adding info that is specific to CATcher.
```suggestion
The Electron framework launches a main process and a renderer process.
The CATcher app runs within the renderer process, while the main process responds to system events that occur during the app's lifecycle. Electron provides a message-passing API to facilitate inter-process communication.
```

anubh-v commented on dg/tools-frameworks.md:
```diff
@@ -0,0 +1,141 @@
+
+### Resources for Electron
+
+1. [Official Electron Guide](https://www.electronjs.org/docs/tutorial) : This is the official Electron documentation]
```
```suggestion
1. [Official Electron Guide](https://www.electronjs.org/docs/tutorial) : This is the official Electron documentation
```

anubh-v reviewed (commented):
Thanks for the PR. LGTM.

I have some final comments. We can fix these comments in a future PR, if you do not have time now.

kkangs0226 commented on dg/tools-frameworks.md:
````diff
@@ -0,0 +1,141 @@
+```
+This starts the authentication flow in the CATcher desktop version. `ipcMain` communicates with the renderer processes by waiting for events from them, as well as emitting events to them.
+
+Because it is a desktop application, it is important that we account for different operating systems in our code, since different operating systems would behave differently to changes in our application.
````
Alright, got it!

kkangs0226 commented:
> Thanks for the PR. LGTM.
> 
> I have some final comments. We can fix these comments in a future PR, if you do not have time now.

I have made the required changes!
